crates/rpc/rpc/src/eth/filter.rs

```diff
@@ -1014,89 +1023,144 @@ struct RangeBlockMode<
+                        expected = expected_next,
+                        "Non-consecutive block detected, stopping range collection"
+                    );
+                    break; // Non-consecutive block, stop here
+                }
 
-        // Collect consecutive blocks up to max_range size
-        while range_headers.len() < self.max_range {
-            let Some(peeked) = self.iter.peek() else { break };
-            let Some(last_header) = range_headers.last() else { break };
+                let Some(next_header) = self.iter.next() else { break };
+                range_headers.push(next_header);
+            }
 
-            let expected_next = last_header.header().number() + 1;
-            if peeked.header().number() != expected_next {
-                break; // Non-consecutive block, stop here
+            // Check if we should use parallel processing for large ranges
+            let remaining_headers = self.iter.len() + range_headers.len();
+            if remaining_headers >= PARALLEL_PROCESSING_THRESHOLD {
+                self.spawn_parallel_tasks(range_headers);
+            } else {
+                self.spawn_sequential_task(range_headers);
             }
 
-            let Some(next_header) = self.iter.next() else { break };
-            range_headers.push(next_header);
+            // Now wait for the newly spawned task to complete
+            if let Some(task_result) = self.pending_tasks.next().await {
+                let mut chunk_results = task_result?;
+                if !chunk_results.is_empty() {
+                    let first_result = chunk_results.remove(0);
+                    self.next.extend(chunk_results);
+                    return Ok(Some(first_result));
+                }
+            }
         }
 
-        // Check if we should use parallel processing for large ranges
-        let remaining_headers = self.iter.len() + range_headers.len();
-        if remaining_headers >= PARALLEL_PROCESSING_THRESHOLD as usize {
-            self.process_large_range(range_headers).await
-        } else {
-            self.process_small_range(range_headers).await
-        }
+        // If we reach here, a task completed with empty results but other tasks are still pending.
+        // Return None and let the next call handle remaining tasks.
+        debug!(
+            target: "rpc::eth::filter",
+            pending_tasks_len = self.pending_tasks.len(),
+            queued_results_len = self.next.len(),
+            headers_remaining = self.iter.len(),
+            "Task returned empty results with pending tasks remaining - returning None"
+        );
+        Ok(None)
     }
 
-    /// Process small range headers
-    async fn process_small_range(
+    /// Spawn a sequential task for processing a small range of headers
+    ///
+    /// This is used when the remaining headers count is below [`PARALLEL_PROCESSING_THRESHOLD`].
+    fn spawn_sequential_task(
```

Comment:
this doesn't actually spawn a task

I'd like to keep this function at this was previously because this behaviour didn't change, and we can instead await this call directly